ToolBox/CoinApi/CoinApiDataQueueHandler.cs

```diff
@@ -0,0 +1,360 @@
+        {
+            _connectionHandler.ConnectionLost += OnConnectionLost;
+            _connectionHandler.ConnectionRestored += OnConnectionRestored;
+            _connectionHandler.ReconnectRequested += OnReconnectRequested;
+
+            _connectionHandler.Initialize();
+
+            _webSocket.Initialize(WebSocketUrl);
+
+            _webSocket.Message += OnMessage;
+
+            _webSocket.Connect();
+        }
+
+        /// <summary>
+        /// Get the next ticks from the live trading data queue
+        /// </summary>
+        /// <returns>IEnumerable list of ticks since the last update.</returns>
+        public IEnumerable<BaseData> GetNextTicks()
+        {
+            lock (_locker)
+            {
+                var copy = _ticks.ToArray();
+                _ticks.Clear();
+                return copy;
+            }
+        }
+
+        /// <summary>
+        /// Adds the specified symbols to the subscription
+        /// </summary>
+        /// <param name="job">Job we're subscribing for:</param>
+        /// <param name="symbols">The symbols to be added keyed by SecurityType</param>
+        public void Subscribe(LiveNodePacket job, IEnumerable<Symbol> symbols)
+        {
+            lock (_lockerSubscriptions)
+            {
+                var symbolsToSubscribe = (from symbol in symbols
+                                          where !_subscribedSymbols.Contains(symbol) && CanSubscribe(symbol)
+                                          select symbol).ToList();
+                if (symbolsToSubscribe.Count == 0)
+                    return;
+
+                Log.Trace($"CoinApiDataQueueHandler.Subscribe(): {string.Join(",", symbolsToSubscribe.Select(x => x.Value))}");
+
+                // CoinAPI requires at least 5 seconds between subscription requests so we need to batch them
+                _subscribedSymbols = symbolsToSubscribe.Union(_subscribedSymbols).ToHashSet();
+
+                ProcessSubscriptionRequest();
+            }
+        }
+
+        /// <summary>
+        /// Removes the specified symbols to the subscription
+        /// </summary>
+        /// <param name="job">Job we're processing.</param>
+        /// <param name="symbols">The symbols to be removed keyed by SecurityType</param>
+        public void Unsubscribe(LiveNodePacket job, IEnumerable<Symbol> symbols)
+        {
+            lock (_lockerSubscriptions)
+            {
+                var symbolsToUnsubscribe = (from symbol in symbols
+                                            where _subscribedSymbols.Contains(symbol)
+                                            select symbol).ToList();
+                if (symbolsToUnsubscribe.Count == 0)
+                    return;
+
+                Log.Trace($"CoinApiDataQueueHandler.Unsubscribe(): {string.Join(",", symbolsToUnsubscribe.Select(x => x.Value))}");
+
+                // CoinAPI requires at least 5 seconds between subscription requests so we need to batch them
```

Comment:
Yeah, the comment is confusing 😃.
I'll try to explain the issue better:
CoinAPI web sockets API only have one Hello message (where we need to pass all the symbols/markets to be subscribed): https://docs.coinapi.io/#hello

I quickly found out that if you send multiple Hello message quickly, they will simply stop sending data with no error.
I was able to get this message only a couple of times and this is where my 5-second comment comes from:
```{"message":"Between resubscriptions (hello messages) must pass minimum 5 seconds.","type":"error"}```

So we cannot send a Hello message for each Subscribe request but we must group all the symbols together and submit a single Hello message for all symbols.
This is where the 250ms `_subscribeDelay` comes into play: when we don't receive a new subscribe message for 250ms, we assume that the list of symbols is complete and we can submit the request.

A similar logic was implemented for subscribe requests in the Oanda brokerage, although for a different reason.

Reply:
Cool, I get it now 👍 .
The only comment is that after the initial batch of subscriptions are sent, I believe the implementation is not guaranteeing/waiting the 5 seconds if a new subscription request is sent after `_subscribeDelay`, right? Maybe it could keep the state somehow

Reply:
Seems a bit of an edge case, like an algorithm playing with `AddSecurity` and `RemoveSecurity` at Tick or Second resolutions, but should be handled anyway. I'll track the time of the last Hello message and guarantee at least 5 seconds between requests.